modules/styling/lib/cs.ts

```diff
@@ -425,9 +453,75 @@ export function handleCsProp<
+  // _does_ care because it uses the `classList` order to determine style merging. Therefore we
+  // should always be careful to order class names knowing we need to support runtime merging if
+  // runtime styles are detected.
+  const returnProps = csToProps([localCs, className, cs, style]);
+
+  // The following is basically what Emotion does internally to merge styles from different
+  // components into a single className. We will do this here for backwards compatibility with
+  // runtime style merging. `createStyles` and `@emotion/css` works by creating styles and injecting
+  // them immediately into the document. `@emotion/react` and `@emotion/styled` work by injecting
+  // styles during the React render cycle. The following links are to source code of
+  // `@emotion/styled` and `@emotion/react` respectively that merge styles in a similar manner.
+  // - https://github.com/emotion-js/emotion/blob/f3b268f7c52103979402da919c9c0dd3f9e0e189/packages/styled/src/base.js#L120C51-L138
+  // - https://github.com/emotion-js/emotion/blob/f3b268f7c52103979402da919c9c0dd3f9e0e189/packages/react/src/emotion-element.js#L102-L116
+  (returnProps.className || '').split(' ').forEach(name => {
+    // Here we detect if a className is associated with Emotion's cache. Styles created via
+    // `createStyles` do not need special runtime merge treatment, but `@emotion/react` and
+    // `@emotion/styled` do. The `createStylesCache` tracks all styles injected into the cache via
+    // `createStyles`, so those are filtered out.
+    if (!createStylesCache[name] && cache.registered[name]) {
+      shouldRuntimeMergeStyles = true;
+    }
+  });
+
+  // If runtime styles are detected, we need to do extra work to ensure styles merge according to
+  // the rules of Emotion's runtime.
+  if (shouldRuntimeMergeStyles) {
+    const registeredStyles: string[] = [];
+
+    // We are using the raw `css` instead of `createStyles` because `css` uses style hashing and
+    // `createStyles` generates a new ID every time. We could use `createStyles` here, but it would
+    // be more wasteful and new styles would be generated each React render cycle. `css` is safe to
+    // use inside a render function while `createStyles` should always be used outside the React
+    // render cycle. The following is `@emotion/utils` and it mutates the passed in
+    // `registeredStyles` array. It uses the cache of registered styles and the original className.
+    // It returns a string with found registered class names removed.
+    returnProps.className = getRegisteredStyles(
+      cache.registered,
+      registeredStyles,
+      returnProps.className || ''
+    );
+
+    // The `className` is mutated again, but with a brand new CSS class name of all the merged
+    // registered styles. This is how Emotion merges styles from the `css` prop and `styled`
+    // components.
+    returnProps.className += css(registeredStyles);
+  }
```

Comment:
This is the same implementation as what was in `mergeStyles`, but I added a lot more comments to it.